keps/sig-api-machinery/5073-declarative-validation-with-validation-gen/README.md

```diff
@@ -944,23 +948,77 @@ The linter, as previously described, will enforce rules to address valid zero-va
+* An update via `pods/resize` is validated using the standard Pod rules. However, a rule on `spec.container[*].resources` might look like `+k8s:if('subresource != "/resize"')=+k8s:immutable`, effectively enforcing immutability *unless* the update comes via the `resize` subresource.
+
+**Support required:**
+
+* To enable conditional validation, declarative validation provides access to the `subresources` parameter within validation rule expressions.
+
+#### Scale-Type Subresources
+
+These subresources have the following characteristics:
+
+* They often represent a different API `kind` (e.g., `autoscaling/v1.Scale`) than the resource they modify (e.g., `apps/v1.Deployment`).
+* Despite being a different `kind`, updates to the subresource modify fields within the underlying storage object (same object in etcd) as the *primary* resource.
+
+**Examples:**
+
+* An update to `deployments/scale` modifies the `spec.replicas` field within the primary `deployments` resource.
+* Creating a `Binding` object via `pods/binding` sets the `spec.nodeName` field on the primary `pods` resource.
+
+**Validation Process:**
+
+1.  **Subresource Validation:** The incoming subresource object itself (e.g., the `autoscaling/v1.Scale` object) is validated using *its own* declarative rules.
+2.  **Storage Layer Application:** The resource's storage layer logic translates the validated subresource update into changes on the primary resource's fields (e.g., mapping the `Scale` object's `spec.replicas` to the `Deployment` object's `spec.replicas`).
+3.  **Primary Resource Validation:** The *modified primary resource* (e.g., `Deployment`) is then validated using *its* standard declarative validation rules.
+4.  **Ratcheting:** Ratcheting is applied during the primary resource validation, skipping checks on fields that were not affected by the subresource update.
+
+**Support required:**
+
+* Declarative validation will need to provide a easy way for a storage layer to map the internal type of a subresource to the
+  requested versioned type of that resource.  For primary resource validation, the information is present in the requestInfo of
+  the context, but for these validations, the storage layer typically [manages a mapping](https://github.com/kubernetes/kubernetes/blob/30469e180361d7da07b0fee6d47c776fa2cf3e86/pkg/registry/core/replicationcontroller/storage/storage.go#L170-L177) which will need to
+  be used. https://github.com/jpbetz/kubernetes/pull/141 provides an example of migrating a `/scale` subresource and 
+  introduces utilities for managing the subresource mapping.
+
+#### Streaming Subresources
+
+Subresources such as `pods/exec`, `pods/attach`, and `pods/portforward` do not require declarative validation, as they operate on data streams, not structured resource data.
```

Comment:
Permanently or temporarily?  I can see value in validation options for these, though I concede that a first pass doesn't have to provide that.

Reply:
I'd forgotten about the options structs when writing this.  Those are supportable. I'll make note.